pilot/pkg/config/kube/crdclient/client_test.go

```diff
@@ -327,6 +331,75 @@ func TestClient(t *testing.T) {
 	})
 }
 
+// TestClientInitialSyncSkipsOtherRevisions tests that the initial sync skips objects from other
+// revisions.
+func TestClientInitialSyncSkipsOtherRevisions(t *testing.T) {
```

Comment:
getting some flaky tests on this:

```
=== RUN   TestClientInitialSyncSkipsOtherRevisions
2022-12-05T22:08:34.949555Z     warn    kube    Skipping CRD gateway.networking.k8s.io/v1alpha2/ReferenceGrant as it is not present     controller=
2022-12-05T22:08:34.949705Z     warn    kube    Skipping CRD gateway.networking.k8s.io/v1alpha2/TCPRoute as it is not present   controller=
2022-12-05T22:08:34.949777Z     warn    kube    Skipping CRD gateway.networking.k8s.io/v1alpha2/TLSRoute as it is not present   controller=
2022-12-05T22:08:34.949852Z     warn    kube    Skipping CRD gateway.networking.k8s.io/v1beta1/GatewayClass as it is not present        controller=
2022-12-05T22:08:34.949901Z     warn    kube    Skipping CRD gateway.networking.k8s.io/v1beta1/Gateway as it is not present     controller=
2022-12-05T22:08:34.949933Z     warn    kube    Skipping CRD gateway.networking.k8s.io/v1beta1/HTTPRoute as it is not present   controller=
2022-12-05T22:08:34.950097Z     info    kube    Starting Pilot K8S CRD controller       controller=
2022-12-05T22:08:34.950304Z     info    kube    controller "security.istio.io/v1beta1/AuthorizationPolicy" is syncing...        controller=
2022-12-05T22:08:34.952901Z     info    kube    controller "networking.istio.io/v1alpha3/WorkloadGroup" is syncing...   controller=
2022-12-05T22:08:34.961335Z     info    kube    controller "telemetry.istio.io/v1alpha1/Telemetry" is syncing...        controller=
2022-12-05T22:08:34.978650Z     info    kube    Pilot K8S CRD controller synced in 28.576186ms  controller=
    client_test.go:399: expected configs [{{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-0 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-3 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-6 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  }] do not match those added [{{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-0 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-3 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-3 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-6 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-6 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  }]
--- FAIL: TestClientInitialSyncSkipsOtherRevisions (0.05s)
FAIL
```

```
==================
WARNING: DATA RACE
Read at 0x00c000d222e8 by goroutine 57:
  reflect.typedmemmove()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/runtime/mbarrier.go:178 +0x0
  reflect.packEface()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/reflect/value.go:123 +0xaf
  reflect.valueInterface()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/reflect/value.go:1508 +0x18b
  reflect.Value.Interface()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/reflect/value.go:1478 +0xc9
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:753 +0xca
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:837 +0x1db5
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:837 +0x1db5
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:896 +0x2425
  fmt.(*pp).printArg()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:743 +0xe17
  fmt.(*pp).doPrintf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:1057 +0x499
  fmt.Sprintf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:219 +0x67
  testing.(*common).Fatalf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:954 +0x8a
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:399 +0xe1e
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x47

Previous write at 0x00c000d222e8 by goroutine 58:
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions.func1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:380 +0x104
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*cacheHandler).callHandlers()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:97 +0x282
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*cacheHandler).onEvent()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:76 +0xc27
  istio.io/istio/pilot/pkg/config/kube/crdclient.createCacheHandler.func2.1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:125 +0x53
  istio.io/istio/pkg/queue.(*queueImpl).processNextItem()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/queue/instance.go:109 +0x5c
  istio.io/istio/pkg/queue.(*queueImpl).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/queue/instance.go:135 +0x20d
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*Client).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client.go:260 +0x344
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions.func3()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:386 +0x47

Goroutine 57 (running) created at:
  testing.(*T).Run()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x75d
  testing.runTests.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1846 +0x99
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.runTests()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1844 +0x7ec
  testing.(*M).Run()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1726 +0xa84
  istio.io/istio/tests/util/leak.CheckMain()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/tests/util/leak/check.go:151 +0x3b
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestMain()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/leak_test.go:25 +0x32b
  main.main()
      _testmain.go:59 +0x31f

Goroutine 58 (running) created at:
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:386 +0xbe9
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x47
==================
==================
WARNING: DATA RACE
Read at 0x00c000d22328 by goroutine 57:
  reflect.packEface()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/reflect/value.go:130 +0xfd
  reflect.valueInterface()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/reflect/value.go:1508 +0x18b
  reflect.Value.Interface()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/reflect/value.go:1478 +0xc9
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:753 +0xca
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:837 +0x1db5
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:837 +0x1db5
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:896 +0x2425
  fmt.(*pp).printArg()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:743 +0xe17
  fmt.(*pp).doPrintf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:1057 +0x499
  fmt.Sprintf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:219 +0x67
  testing.(*common).Fatalf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:954 +0x8a
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:399 +0xe1e
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x47

Previous write at 0x00c000d22328 by goroutine 58:
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions.func1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:380 +0x104
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*cacheHandler).callHandlers()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:97 +0x282
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*cacheHandler).onEvent()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:76 +0xc27
  istio.io/istio/pilot/pkg/config/kube/crdclient.createCacheHandler.func2.1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:125 +0x53
  istio.io/istio/pkg/queue.(*queueImpl).processNextItem()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/queue/instance.go:109 +0x5c
  istio.io/istio/pkg/queue.(*queueImpl).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/queue/instance.go:135 +0x20d
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*Client).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client.go:260 +0x344
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions.func3()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:386 +0x47

Goroutine 57 (running) created at:
  testing.(*T).Run()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x75d
  testing.runTests.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1846 +0x99
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.runTests()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1844 +0x7ec
  testing.(*M).Run()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1726 +0xa84
  istio.io/istio/tests/util/leak.CheckMain()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/tests/util/leak/check.go:151 +0x3b
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestMain()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/leak_test.go:25 +0x32b
  main.main()
      _testmain.go:59 +0x31f

Goroutine 58 (running) created at:
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:386 +0xbe9
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x47
==================
==================
WARNING: DATA RACE
Read at 0x00c000d22368 by goroutine 57:
  reflect.Value.Len()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/reflect/value.go:1617 +0x247d
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:892 +0x2426
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:837 +0x1db5
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:837 +0x1db5
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:896 +0x2425
  fmt.(*pp).printArg()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:743 +0xe17
  fmt.(*pp).doPrintf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:1057 +0x499
  fmt.Sprintf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:219 +0x67
  testing.(*common).Fatalf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:954 +0x8a
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:399 +0xe1e
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x47

Previous write at 0x00c000d22368 by goroutine 58:
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions.func1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:380 +0x104
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*cacheHandler).callHandlers()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:97 +0x282
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*cacheHandler).onEvent()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:76 +0xc27
  istio.io/istio/pilot/pkg/config/kube/crdclient.createCacheHandler.func2.1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:125 +0x53
  istio.io/istio/pkg/queue.(*queueImpl).processNextItem()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/queue/instance.go:109 +0x5c
  istio.io/istio/pkg/queue.(*queueImpl).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/queue/instance.go:135 +0x20d
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*Client).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client.go:260 +0x344
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions.func3()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:386 +0x47

Goroutine 57 (running) created at:
  testing.(*T).Run()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x75d
  testing.runTests.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1846 +0x99
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.runTests()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1844 +0x7ec
  testing.(*M).Run()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1726 +0xa84
  istio.io/istio/tests/util/leak.CheckMain()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/tests/util/leak/check.go:151 +0x3b
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestMain()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/leak_test.go:25 +0x32b
  main.main()
      _testmain.go:59 +0x31f

Goroutine 58 (running) created at:
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:386 +0xbe9
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x47
==================
==================
WARNING: DATA RACE
Read at 0x00c000d22380 by goroutine 57:
  reflect.Value.IsNil()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/reflect/value.go:1552 +0x99
  fmt.getField()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:323 +0x65
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:837 +0x1d89
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:896 +0x2425
  fmt.(*pp).printArg()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:743 +0xe17
  fmt.(*pp).doPrintf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:1057 +0x499
  fmt.Sprintf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:219 +0x67
  testing.(*common).Fatalf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:954 +0x8a
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:399 +0xe1e
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x47

Previous write at 0x00c000d22380 by goroutine 58:
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions.func1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:380 +0x104
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*cacheHandler).callHandlers()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:97 +0x282
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*cacheHandler).onEvent()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:76 +0xc27
  istio.io/istio/pilot/pkg/config/kube/crdclient.createCacheHandler.func2.1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:125 +0x53
  istio.io/istio/pkg/queue.(*queueImpl).processNextItem()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/queue/instance.go:109 +0x5c
  istio.io/istio/pkg/queue.(*queueImpl).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/queue/instance.go:135 +0x20d
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*Client).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client.go:260 +0x344
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions.func3()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:386 +0x47

Goroutine 57 (running) created at:
  testing.(*T).Run()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x75d
  testing.runTests.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1846 +0x99
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.runTests()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1844 +0x7ec
  testing.(*M).Run()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1726 +0xa84
  istio.io/istio/tests/util/leak.CheckMain()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/tests/util/leak/check.go:151 +0x3b
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestMain()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/leak_test.go:25 +0x32b
  main.main()
      _testmain.go:59 +0x31f

Goroutine 58 (running) created at:
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:386 +0xbe9
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x47
==================
    client_test.go:399: expected configs [{{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-0 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-3 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-6 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  }] do not match those added [{{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-0 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-3 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-6 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-6 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  }]
    testing.go:1319: race detected during execution of test
--- FAIL: TestClientInitialSyncSkipsOtherRevisions (0.06s)
=== CONT
    testing.go:1319: race detected during execution of test
FAIL
```

Reply:
Thanks for finding that, that race was incredibly rare on my machine had to run the tests 20,000 times to get it to reliably repro.

I've fixed it by calling `SyncAll()` directly to do the initial bootstrap sync. I tried closing the stop channel and waiting for the store to finish, but that led to another rare flake where the k8s fake client would occasionally send extra add events that the store would then pick up, so calling `SyncAll()` appears to be the safest option.

```
% go test -race -count=20000 ./pilot/pkg/config/kube/crdclient/ -test.run=TestClientInitialSyncSkipsOtherRevisions
ok  	istio.io/istio/pilot/pkg/config/kube/crdclient	75.443s
```